### PR TITLE
feat: refactor chains & assets

### PR DESCRIPTION
**Pull request by kraikov**
This PR contains only logic for a single EVM Chain & single UTXO chain.
Once discussed and approved, the rest of the chains will be included

### CODE REVIEW

File: packages/cryptoassets/src/chains/BaseChain.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import { IAsset } from '../interfaces/IAsset';
+import { IChain } from '../interfaces/IChain';
+import { IFees } from '../interfaces/IFees';
+import { IGasLimits } from '../interfaces/IGasLimit';
+import { INetwork } from '../interfaces/INetwork';
+import { ChainId } from '../types';
+import { ExplorerView, FeeMultiplier } from './types';
+
+export abstract class BaseChain implements IChain {
+  constructor(chain: {
```
Comment (monokh) [inline packages/cryptoassets/src/chains/BaseChain.ts]: I think you could skip a bunch of lines by making `IChain` only have the properties 
Then accepting IChain in the constructor and doing `Object.assign(this, chain)`

Not sure if that will work for everything else you are doing

Comment (kraikov) [reply packages/cryptoassets/src/chains/BaseChain.ts]: Good idea

File: packages/cryptoassets/src/chains/EvmChains.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import { IChain } from '../interfaces/IChain';
+import { AssetTypes, ChainId } from '../types';
+import { EvmChain } from './EvmChain';
+
+export const EVM_CHAINS: { [key in ChainId]?: IChain } = {
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/EvmChains.ts]: Declare a type for `{ [key in ChainId]?: IChain }`.

File: packages/cryptoassets/src/interfaces/IFees.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export interface IFees {
+  unit: string;
+  magnitute: number;
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/interfaces/IFees.ts]: Spelling? Did you mean `magnitude`?

File: packages/cryptoassets/src/chains/BaseChain.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  public multicallSupport: boolean;
+
+  public EIP1559: boolean;
+  public feeMultiplier: FeeMultiplier;
+  public gasLimit: IGasLimits;
+  public fees: IFees;
+  public supportCustomFees: boolean;
+
+  public isValidAddress(address: string) {
+    return isAddress(address);
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/BaseChain.ts]: Why do we use EVM methods from `@ethersproject/address` in the base class?
Maybe `isValidAddress` should be abstract?

File: packages/cryptoassets/src/chains/BaseChain.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  public formatAddress(address: string) {
+    return getAddress(address);
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/BaseChain.ts]: Why do we use EVM methods from `@ethersproject/address` in the base class?
Maybe `formatAddress` should be abstract?

File: packages/cryptoassets/src/chains/BaseChain.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  public isValidTransactionHash(hash: string) {
+    return /^(0x)?([A-Fa-f0-9]{64})$/.test(hash);
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/BaseChain.ts]: This is check works only for EVM chains.

File: packages/cryptoassets/src/chains/index.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import { EVM_CHAINS } from './EvmChains';
+import { UTXO_CHAINS } from './UtxoChains';
+
+export const chains = { ...EVM_CHAINS, ...UTXO_CHAINS };
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/index.ts]: Why not simply export а getter (with checks for existence and etc.) and maybe a setter method.

File: packages/cryptoassets/src/chains/BaseChain.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import { IGasLimits } from '../interfaces/IGasLimit';
+import { INetwork } from '../interfaces/INetwork';
+import { ChainId, ExplorerView, FeeMultiplier } from '../types';
+
+export abstract class BaseChain implements IChain {
+  constructor(chain: IChain) {
+    Object.assign(this, chain);
+  }
+
+  public id: ChainId;
```
Comment (monokh) [inline packages/cryptoassets/src/chains/BaseChain.ts]: I think you can avoid defining all the properties in the class like this:

```
export interface BaseChain implements IChain

export abstract class BaseChain implements IChain {
...
```
Don't ask me how this works 😄

File: packages/cryptoassets/src/chains/mainnet/EvmChains.ts [hunk 9]
```diff
@@ -0,0 +1,410 @@
+import { IChain } from '../../interfaces/IChain';
+import { AssetTypes, ChainId } from '../../types';
+import { EvmChain, RskChain } from '../EvmChain';
+
+export const EVM_CHAINS: { [key in ChainId]?: IChain } = {
+  [ChainId.Ethereum]: new EvmChain({
```
Comment (monokh) [inline packages/cryptoassets/src/chains/mainnet/EvmChains.ts]: nice to put chain configs each in its own file?

**Review by monokh — APPROVED** [review]

File: packages/cryptoassets/src/assets/testnet/native.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import { TESTNET_SUPPORTED_CHAINS } from '../../chains';
+import { IChain } from '../../interfaces/IChain';
+import { AssetMap } from '../../types';
+
+const TESTNET_NATIVE_ASSETS = Object.values(TESTNET_SUPPORTED_CHAINS).reduce((result: AssetMap, chain: IChain) => {
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/assets/testnet/native.ts:5]: Same code as `mainnet/native.ts`

File: packages/cryptoassets/src/chains/mainnet/evm/arbitrum.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  safeConfirmations: 5,
+  txFailureTimeoutMs: 600_000,
+
+  network: {
+    name: 'arbitrum_mainnet',
+    coinType: '60',
+    networkId: 42161,
+    chainId: 42161,
+    isTestnet: false,
+    rpcUrls: ['https://arbitrum-mainnet.infura.io/v3/da99ebc8c0964bb8bb757b6f8cc40f1f'],
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/mainnet/evm/arbitrum.ts:37]: Is it possible to read infura Key from `.env`.

File: packages/cryptoassets/src/chains/mainnet/evm/avalanche.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  txFailureTimeoutMs: 600_000,
+
+  network: {
+    name: 'avalanche_mainnet',
+    coinType: '60',
+    networkId: 43114,
+    chainId: 43114,
+    isTestnet: false,
+    rpcUrls: [
+      'https://speedy-nodes-nyc.moralis.io/7c28a10f7d39bfb24704dafc/avalanche/mainnet',
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/mainnet/evm/avalanche.ts:37]: Is it possible to read moralis key from `.env`

File: packages/cryptoassets/src/chains/mainnet/evm/ethereum.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  txFailureTimeoutMs: 3_600_000, // 1 hour
+
+  network: {
+    name: 'ethereum_mainnet',
+    coinType: '60',
+    isTestnet: false,
+    chainId: 1,
+    networkId: 1,
+    rpcUrls: [
+      'https://mainnet.infura.io/v3/da99ebc8c0964bb8bb757b6f8cc40f1f',
```
Comment (skeremidchiev) [inline packages/cryptoassets/src/chains/mainnet/evm/ethereum.ts:36]: Same issue with keys.

File: packages/wallet-core/src/swaps/oneinch/OneinchSwapProvider.ts [hunk 14]
```diff
@@ -96,7 +96,8 @@ class OneinchSwapProvider extends SwapProvider {
   }
 
   async getQuote({ network, from, to, amount }: QuoteRequest) {
-    if (!isEthereumChain(from) || !isEthereumChain(to) || new BigNumber(amount).lte(0)) return null;
+    if (!isChainEvmCompatible(from, network) || !isChainEvmCompatible(to, network) || new BigNumber(amount).lte(0))
+      return null;
```
Comment (skeremidchiev) [inline packages/wallet-core/src/swaps/oneinch/OneinchSwapProvider.ts:93]: `if {}` !

**Review by AndonMitev — APPROVED** [review]
LGTM LIT FIRE

**Review by skeremidchiev — APPROVED** [review]

File: packages/cryptoassets/src/chains/index.ts (hunk 7 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/index.ts]: There is one.

File: packages/cryptoassets/src/assets/testnet/native.ts (hunk 10 above)
Comment (kraikov) [reply packages/cryptoassets/src/assets/testnet/native.ts:5]: There is one.

File: packages/cryptoassets/src/chains/mainnet/evm/arbitrum.ts (hunk 11 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/mainnet/evm/arbitrum.ts:37]: no, cryptoassets cannot have env variables

File: packages/cryptoassets/src/chains/mainnet/evm/avalanche.ts (hunk 12 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/mainnet/evm/avalanche.ts:37]: no, cryptoassets cannot have env variables

File: packages/cryptoassets/src/chains/mainnet/evm/ethereum.ts (hunk 13 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/mainnet/evm/ethereum.ts:36]: no, cryptoassets cannot have env variables

File: packages/wallet-core/src/swaps/oneinch/OneinchSwapProvider.ts (hunk 14 above)
Comment (kraikov) [reply packages/wallet-core/src/swaps/oneinch/OneinchSwapProvider.ts:93]: i do agree, but not part of this PR 🗡️

File: packages/cryptoassets/src/chains/BaseChain.ts (hunk 4 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/BaseChain.ts]: That's outdated file?

File: packages/cryptoassets/src/chains/BaseChain.ts (hunk 5 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/BaseChain.ts]: That's outdated file?

File: packages/cryptoassets/src/chains/BaseChain.ts (hunk 6 above)
Comment (kraikov) [reply packages/cryptoassets/src/chains/BaseChain.ts]: That's outdated file?

**Review by monokh — APPROVED** [review]

**Pull request merged**